Clean up container rootf mounts on daemon start, fixes #19679

coolljt0725 opened the pull request:
When the daemon shutdown ungracefull, it will left the running
containers' rootfs still be mounted. This will cause some error
when trying to remove the containers.

<del>and remove 
`daemon.Unmount(container)` from `Register` since it's duplicate 
with https://github.com/docker/docker/blob/master/daemon/daemon.go#L805</del>

This is for overlay graphdriver

ping @cpuguy83 
/cc @ibuildthecloud 

Signed-off-by: Lei Jitang leijitang@huawei.com

calavera commented:
This seems like a good approach, thanks @coolljt0725. I'm wondering what will happen with windows if we remove that unmount call. It might be okay, but I'd like to hear from @jhowardmsft first.

jhowardmsft commented:
@swernli Would welcome a second opinion. I'm pretty sure this is safe for Windows, but best get more eyes on it. 

PS gccgo Jenkins failed - kicked off again

coolljt0725 commented:
@calavera  @jhowardmsft  After checking, I prefer not remove the `daemon.Unmount(container)` ,
`d.cleanupMounts()` only clean the mounts of `overlay`. WDYT?

swernli commented:
@coolljt0725 @jhowardmsft Yes, I agree with the decision to leave the unmounts call as is.  Removing it would be problematic, but the most recent version of the changes keep the expected behavior for Windows.  Thanks!

calavera commented:
LGTM

cpuguy83 commented on daemon/daemon_linux.go:
```diff
@@ -33,11 +33,11 @@ func (daemon *Daemon) cleanupMountsFromReader(reader io.Reader, unmount func(tar
 			mnt := fields[4]
 			mountBase := filepath.Base(mnt)
-			if mountBase == "mqueue" || mountBase == "shm" {
+			if mountBase == "mqueue" || mountBase == "shm" || mountBase == "merged" {
```
I'm not sure I like this.
Cleanup is tied to overlay. Wouldn't other graphdrivers have this issue as well?

coolljt0725 replied:
I tested `overlay`, `devicemapper`, `aufs` and `overlay` `devicemapper` have this issue. 
Yes, this should also for `devicemapper`

coolljt0725 commented:
@cpuguy83  @calavera  update the fix to also handle `devicemapper`

LK4D4 commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -2070,3 +2070,20 @@ func (s *DockerDaemonSuite) TestRunLinksChanged(c *check.C) {
 	c.Assert(out, check.Not(checker.Contains), "1 packets transmitted, 1 packets received")
 }
+
+func (s *DockerDaemonSuite) TestDaemonUngracefulShutdownCleanUpRootfsOfOverlay(c *check.C) {
```
We have a more generic test about cleaning up: `TestCleanupMountsAfterCrash`. Do you think it might be enough? Or what we testing here?

coolljt0725 replied:
@LK4D4  The test `TestCleanupMountsAfterCrash` is based on `vfs` graphdriver. This test is for `overlay`

LK4D4 replied:
No, that test for any graphdriver. All daemons will be run with `--storage-driver=overlay` if `DOCKER_GRAPHDRIVER=overlay` is set. Your test will be run only in that case, too.

coolljt0725 replied:
@LK4D4 I didn't make it clear on my last post,  I mean this test will still work even if we don't set `DOCKER_GRAPHDRIVER=overlay`. The CI is use `vfs` by default, so there is no need to trigger another
CI with another graphdriver.

LK4D4 replied:
I don't understand how it works without it if it requires `/` to be mounted as overlay. It won't be run on CI without `DOCKER_GRAPHDRIVER=overlay`. Also, it did not test that there are no leftovers. Only that container was removed successfully.

coolljt0725 replied:
Oh, I just think the CI environment is using `overlay`. yes, it require `/` to be mounted as overlay. why check if the container was removed successful is related the issue #19679. I'm going to remove this test case since `TestCleanupMountsAfterCrash` is enough.

coolljt0725 committed:
Clean up container rootf mounts on daemon start fixes #19679

When the daemon shutdown ungracefully, it will left the running
containers' rootfs still be mounted. This will cause some error
when trying to remove the containers.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

coolljt0725 commented:
@LK4D4  @calavera  @cpuguy83  updated

calavera commented:
still LGTM

tiborvass commented:
LGTM
